python-stdlib/tarfile/tarfile/__init__.py

```diff
@@ -63,7 +64,7 @@ def skip(self):
 class TarInfo:
     def __init__(self, name=""):
         self.name = name
-        self.mode = _S_IFDIR if self.name[-1] == "/" else _S_IFREG
+        self.mode = _S_IFDIR if os.stat(name)[0] == 0x4000 else _S_IFREG
```

Comment:
What happens if `name` doesn't exist on disk?  In the case that you are extracting from a tar file, the files most likely do not exist on disk, so this will now raise an exception because the file isn't found.

Reply:
thank you @dpgeorge 
I didn't think of that at all 🤔 

would you be ok if I added a `try/except` condition?
I like the flexibility offered by not having to rely on what the user hands in.

In my code I always test if the fs item exists, but I neglected to do it in this case

Reply:
I dug deeper and probably have to handle archive and expand in two different ways.
During expansion it does need to rely on the name coming from the item in the archive 🤔

Reply:
@dpgeorge @projectgus 
please take a look at my last commit, I think I found a compromise 😄